decompress.c

```diff
@@ -65,6 +65,68 @@ static sqfs_err sqfs_decompressor_xz(void *in, size_t insz,
 	return SQFS_OK;
 }
 #define CAN_DECOMPRESS_XZ 1
+
+// The following implementation was adapted from  squashfs-tools
+// https://github.com/plougher/squashfs-tools.git
+
+#define LZMA_PROPS_SIZE 5
+#define LZMA_UNCOMP_SIZE 8
+#define LZMA_HEADER_SIZE (LZMA_PROPS_SIZE + LZMA_UNCOMP_SIZE)
+#define MEMLIMIT (32 * 1024 * 1024)
+
+static sqfs_err sqfs_decompressor_lzma(void *in, size_t insz,
+		void *out, size_t *outsz) {
+
+	lzma_stream strm = LZMA_STREAM_INIT;
+	int uncompressed_size = 0, res;
+	unsigned char lzma_header[LZMA_HEADER_SIZE];
+
+	res = lzma_alone_decoder(&strm, MEMLIMIT);
+	if(res != LZMA_OK) {
+		lzma_end(&strm);
+		return SQFS_ERR;
+	}
+
+	memcpy(lzma_header, in, LZMA_HEADER_SIZE);
+	uncompressed_size = lzma_header[LZMA_PROPS_SIZE] |
+		(lzma_header[LZMA_PROPS_SIZE + 1] << 8) |
+		(lzma_header[LZMA_PROPS_SIZE + 2] << 16) |
+		(lzma_header[LZMA_PROPS_SIZE + 3] << 24);
+
+	if(uncompressed_size > *outsz) {
+		lzma_end(&strm);
+		return SQFS_ERR;
+	}
+
+	memset(lzma_header + LZMA_PROPS_SIZE, 255, LZMA_UNCOMP_SIZE);
+
+	strm.next_out = out;
+	strm.avail_out = *outsz;
+	strm.next_in = lzma_header;
+	strm.avail_in = LZMA_HEADER_SIZE;
+
+	res = lzma_code(&strm, LZMA_RUN);
+
+	if(res != LZMA_OK || strm.avail_in != 0) {
+		lzma_end(&strm);
+		return SQFS_ERR;
+	}
+
+	strm.next_in = in + LZMA_HEADER_SIZE;
+	strm.avail_in = insz - LZMA_HEADER_SIZE;
+
+	res = lzma_code(&strm, LZMA_FINISH);
+	lzma_end(&strm);
+
+	if(res == LZMA_STREAM_END || (res == LZMA_OK &&
```

Comment:
I don't understand this at all. The various examples in XZ seem to assume we always get LZMA_STREAM_END if we're really done, see: 

https://github.com/xz-mirror/xz/blob/2267f5b0d20a5d24e93fcd9f72ea7eeb0d89708c/doc/examples/02_decompress.c
https://github.com/xz-mirror/xz/blob/2267f5b0d20a5d24e93fcd9f72ea7eeb0d89708c/src/liblzma/common/stream_buffer_decoder.c

Reply:
Maybe I don't get something but the following [here](https://github.com/xz-mirror/xz/blob/2267f5b0d20a5d24e93fcd9f72ea7eeb0d89708c/doc/examples/02_decompress.c#L167) says that we always need to check for the `END`. In our case, the decoding is done in one step for a given bugger so the code just validates the flow as it should be.